weblate/vcs/tests/test_vcs.py

```diff
@@ -435,19 +434,406 @@ def add_remote_commit(self, conflict=False, rename=False):
+
+    def post_side_effect(self, *args, **kwargs):
+        if args[0].find("fork") != -1:
+            return self._mock_response(
+                status=200,
+                json_data={
+                    "ssh_url_to_repo": "git@gitlab.com:test/test.git",
+                    "_links": {"self": "https://gitlab.com/api/v4/projects/20227391"},
+                },
+            )
+        return self._mock_response(
+            status=200,
+            json_data={
+                "web_url": "https://gitlab.com/WeblateOrg/test/-/merge_requests/1"
+            },
+        )
+
+    def get_empty_side_effect(self, *args, **kwargs):
+        if args[0].find("forks") != -1:
+            return self._mock_response(status=200, json_data=[])
+
+        return self._mock_response(status=200, json_data={"id": 20227391})
+
+    def get_side_effect(self, *args, **kwargs):
+        if args[0].find("forks") != -1:
+            return self._mock_response(
+                status=200,
+                json_data=[
+                    {
+                        "ssh_url_to_repo": "git@gitlab.com:test/test.git",
+                        "owner": {"username": "test"},
+                        "_links": {
+                            "self": "https://gitlab.com/api/v4/projects/20227391"
+                        },
+                    }
+                ],
+            )
+        return self._mock_response(status=200, json_data={"id": 20227391})
+
+    def test_api_url(self):
+        self.repo.component.repo = "https://gitlab.com/WeblateOrg/test.git"
+        self.assertEqual(
+            self.repo.api_url(), "https://gitlab.com/api/v4/projects/WeblateOrg%2Ftest"
+        )
+
+    def test_push(self, branch=""):
+        self.repo.component.repo = "https://gitlab.com/WeblateOrg/test.git"
+
+        # Patch push_to_fork() function because we don't want to actually
+        # make a git push request
+        mock_push_to_fork_patcher = patch(
+            "weblate.vcs.git.GitMergeRequestBase.push_to_fork"
+        )
+        mock_push_to_fork = mock_push_to_fork_patcher.start()
+        mock_push_to_fork.return_value = ""
+
+        # Mock post, put and get requests for both the fork and PR requests sent.
+        mock_post_patcher = patch("requests.post")
+        mock_post = mock_post_patcher.start()
+        mock_post.side_effect = self.post_side_effect
+
+        mock_put_patcher = patch("requests.put")
+        mock_put = mock_put_patcher.start()
+        mock_put.return_value = self._mock_response(
+            status=200, json_data={"web_url": "https://gitlab.com/test/test"},
+        )
+
+        mock_get_patcher = patch("requests.get")
+        mock_get = mock_get_patcher.start()
+        mock_get.side_effect = self.get_empty_side_effect
 
-class VCSGerritTest(VCSGithubTest):
+        super().test_push(branch)
+        mock_get_patcher.stop()
+        mock_put_patcher.stop()
+        mock_post_patcher.stop()
+        mock_push_to_fork.stop()
```

Comment:
Can't this use contexts instead of manually stopping all patchers? The manual approach seems quite fragile and easy to miss something.